monitoring/nethealth.go

```diff
@@ -0,0 +1,321 @@
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+
+		// Keep only the last `seriesCapacity` number of data points.
+		if len(series) >= c.SeriesCapacity && len(series) > 0 {
+			series = series[1:]
+		}
+		series = append(series, int64(metric.GetCounter().GetValue()))
+
+		if err := c.setTimeoutSeries(peerName, series); err != nil {
+			errors = append(errors, trace.Wrap(err))
+		}
+
+		// Record updated nodes to be returned for use in later nethealth verification step.
+		updated = append(updated, peerName)
+	}
+	return updated, trace.NewAggregate(errors...)
+}
+
+// verifyNethealth verifies that the network communication is healthy for the
+// nodes specified by the provided list of names.
+func (c *nethealthChecker) verifyNethealth(names []string) error {
+	var errors []error
+	for _, name := range names {
+		series, err := c.getTimeoutSeries(name)
+		if err != nil {
+			errors = append(errors, trace.Wrap(err))
+			continue
+		}
+
+		if !c.isHealthy(series) {
+			errors = append(errors, trace.BadParameter("network communication failure with %s", name))
+		}
+	}
+	return trace.NewAggregate(errors...)
+}
+
+// isHealthy returns false if the number of timeouts increases at each data point.
+func (c *nethealthChecker) isHealthy(series []int64) bool {
+	// Checker has not collected enough data yet to check network health.
+	if len(series) < c.SeriesCapacity {
+		return true
+	}
+
+	// Increase in number of timeouts at each data point indicates network issue.
+	for i := 0; i < c.SeriesCapacity-1; i++ {
+		if series[i] >= series[i+1] {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getTimeoutSeries returns the time series data mapped to the specified name.
+// Returns an empty slice if name was not previously mapped.
+func (c *nethealthChecker) getTimeoutSeries(name string) (series []int64, err error) {
+	c.Lock()
+	defer c.Unlock()
+	if value, ok := c.timeoutStats.Get(name); ok {
+		if series, ok = value.([]int64); !ok {
+			return series, trace.BadParameter("couldn't parse time series as []int64; got type %T", value)
+		}
+	}
+	return series, nil
+}
+
+// setTimeoutSeries maps the name to the timeout series.
+func (c *nethealthChecker) setTimeoutSeries(name string, series []int64) error {
+	c.Lock()
+	defer c.Unlock()
+	return c.timeoutStats.Set(name, series, timeoutStatsTTLSeconds)
+}
+
+// getNethealthAddr returns the address of the local nethealth pod.
+func (c *nethealthChecker) getNethealthAddr() (string, error) {
+	pods, err := c.Client.CoreV1().Pods(nethealthNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// Find nethealth pod with matching host ip address.
+	for _, pod := range pods.Items {
+		if pod.GetLabels()[nethealthLabel] != nethealthValue {
+			continue
+		}
+		if pod.Status.HostIP == c.HostIP {
+			return fmt.Sprintf("http://%s:%d", pod.Status.PodIP, c.NethealthPort), nil
+		}
+	}
+	return "", trace.NotFound("unable to find local nethealth pod")
```

Comment:
Scenario where nethealth pod can't be found should be handled gracefully. Think about the upgrades - a new planet agent comes up but monitoring app still hasn't been upgraded and as such there's no nethealth pod running yet.

I would handle this not found error on the caller side gracefully and maybe just log it but not fail the checker.